### DISCUSSION TITLE
CORS Fairing: Turned missing route OPTIONS /api/v2/auth/register into an OPTIONS pre-flight request

### DISCUSSION THREAD

**Discussion by sLINK1337** [rwf2/Rocket#2506] [Questions]
Hello i got some Problems with my Page, i can't use the Register there is a problem i don't know how i can get it to work.

`"Error: Captcha validation failed, please try again."`

and on the Server i get this Message

`=> Error: No matching routes for OPTIONS /api/v2/auth/register.
=> Warning: Responding with 404 Not Found catcher.
=> CORS Fairing: Turned missing route OPTIONS /api/v2/auth/register into an OPTIONS pre- 
     flight request
=> Response succeeded.
    POST /api/v2/auth/register application/json:
=> Matched: POST /api/v2/auth/register application/json (register)
    sending request POST https://www.google.com/recaptcha/api/siteverify
=> Outcome: Success
=> Response succeeded.`

i tried everything but nothing works, so maybe someone here in the Community can help me with the Problem.

```
#[post("/auth/register", format = "json", data = "<request_data>")]
pub fn register(conn: MainPGDatabase, request_data: Json<RegisterRequest>) -> JsonValue {
    if !verify_recaptcha(&request_data.captcha) {
        return json!({
          "success": false,
          "message": "Captcha validation failed, please try again."
        });
    };

    match account_services::register_user(
        &request_data.email,
        &request_data.username,
        &request_data.password,
        conn,
    ) {
        Ok(data) => {
            return json!({
              "success": true,
              "token": data
            });
        }
        Err(errmessage) => {
            return json!({
              "success": false,
              "message": errmessage
            });
        }
    }
}

fn verify_recaptcha(response: &String) -> bool {
    let data = ureq::post("https://www.google.com/recaptcha/api/siteverify").send_form(&[
        ("secret", &std::env::var("CAPTCHA_KEY").unwrap()),
        ("response", response),
    ]);

    let json = data.into_json().unwrap();

    let success: bool = json.get("success").unwrap().as_bool().unwrap();

    success
}
```

```
        async registerUser() {
      if (this.processingRequest) { return };
      this.registerError = false

      if (this.passwordScore <= 1) {
        this.registerError = 'Password is too weak \n'
        return
      }

      this.signInForm.username = this.registerForm.username
      this.signInForm.password = this.registerForm.password

      if (this.registerForm.password !== this.registerForm.password_confirmation) {
        this.registerError = "Passwords do not match"
        return
      }

      const token = await this.$recaptcha.getResponse()

      this.processingRequest = true

      fetch('http://example.com/api/v2/auth/register', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
              email: this.registerForm.email,
              username: this.registerForm.username,
              password: this.registerForm.password,
              captcha: token
          })
      })
          .then(response => response.json())
          .then(data => this.processResponseRegister(data))
          .catch(err => {
              this.processingRequest = false
              this.registerError = err.message
          });
    },
```

**Comment by domenicogaeni** [comment]
Hi, it sounds like you have a CORS problem. A few days ago I had the same problem as you. To solve this problem you have two choices:
- use this rust crate: https://docs.rs/rocket_cors/latest/rocket_cors/.
- otherwise if you don't want to include any external dependency, you can create a manual fairing and include it. For example you can write this fairing:

    ```
    use rocket::fairing::{Fairing, Info, Kind};
    use rocket::http::{Header, Method, Status};
    use rocket::{Request, Response};
      
    pub struct CORS;

    #[rocket::async_trait]
    impl Fairing for CORS {
        fn info(&self) -> Info {
            Info {
                name: "Add CORS headers to responses",
                kind: Kind::Response,
            }
        }

        async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
            if request.method() == Method::Options {
                response.set_status(Status::NoContent);
                response.set_header(Header::new(
                    "Access-Control-Allow-Methods",
                    "POST, PATCH, GET, DELETE",
                ));
                response.set_header(Header::new(
                    "Access-Control-Allow-Headers",
                    "content-type, authorization",
                ));
            }
            response.set_header(Header::new("Access-Control-Allow-Origin", "*"));
            response.set_header(Header::new("Access-Control-Allow-Credentials", "true"));
            response.set_header(Header::new("Vary", "Origin"));
        }
    }
    ```

and then include it for example in this way:

`rocket::build().attach(CORS)
`

I hope I helped you.

**Comment by sLINK1337** [comment]
> Hi, it sounds like you have a CORS problem. A few days ago I had the same problem as you. To solve this problem you have two choices:
> 
> * use this rust crate: https://docs.rs/rocket_cors/latest/rocket_cors/.
> * otherwise if you don't want to include any external dependency, you can create a manual fairing and include it. For example you can write this fairing:
>   ```
>   use rocket::fairing::{Fairing, Info, Kind};
>   use rocket::http::{Header, Method, Status};
>   use rocket::{Request, Response};
>     
>   pub struct CORS;
>   
>   #[rocket::async_trait]
>   impl Fairing for CORS {
>       fn info(&self) -> Info {
>           Info {
>               name: "Add CORS headers to responses",
>               kind: Kind::Response,
>           }
>       }
>   
>       async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
>           if request.method() == Method::Options {
>               response.set_status(Status::NoContent);
>               response.set_header(Header::new(
>                   "Access-Control-Allow-Methods",
>                   "POST, PATCH, GET, DELETE",
>               ));
>               response.set_header(Header::new(
>                   "Access-Control-Allow-Headers",
>                   "content-type, authorization",
>               ));
>           }
>           response.set_header(Header::new("Access-Control-Allow-Origin", "*"));
>           response.set_header(Header::new("Access-Control-Allow-Credentials", "true"));
>           response.set_header(Header::new("Vary", "Origin"));
>       }
>   }
>   ```
> 
> and then include it for example in this way:
> 
> `rocket::build().attach(CORS) `
> 
> I hope I helped you.

i tried to integrate it but still the same problem

**Comment by robertsarkozi** [comment]
The second approach did help me, also avoiding to include rocket_cors.
Thanks

**Comment by watsom27** [comment]
@sLINK1337 You will need to add a route for the `OPTIONS` request too.

An example from one of my projects (fn body removed for simplicity):

```rust
#[put("/update/<application_handle>")]
fn update(application_handle: String) -> Result<(), Status> {
    [...]
}

#[options("/update/<_application_handle>")]
fn cors_check_options(_application_handle: String) -> Status {
    Status::Ok
}
```

Without the `options` route, the browser will try and send an `options` request to your endpoint, but rocket will reject it with `404` because it fails to find a handler for it.

Hope this helps
